haystack/dataclasses/chat_message.py

```diff
@@ -111,10 +313,23 @@ def to_dict(self) -> Dict[str, Any]:
         :returns:
             Serialized version of the object.
         """
-        data = asdict(self)
-        data["role"] = self.role.value
+        serialized: Dict[str, Any] = {}
+        serialized["_role"] = self._role.value
+        serialized["_meta"] = self._meta
 
-        return data
+        content: List[Dict[str, Any]] = []
+        for part in self._content:
+            if isinstance(part, TextContent):
+                content.append({"text": part.text})
+            elif isinstance(part, ToolCall):
+                content.append({"tool_call": asdict(part)})
+            elif isinstance(part, ToolCallResult):
+                content.append({"tool_call_result": asdict(part)})
+            else:
+                raise TypeError(f"Unsupported type in ChatMessage content: `{type(part).__name__}` for `{part}`.")
+
+        serialized["_content"] = content
+        return serialized
```

Comment:
I still do not like the idea that the serialized chatmessage has underscore prefixes. May I ask what is the rationale behind it?

Reply:
This simply stems from the new dataclass definition:
```python
@dataclass
class ChatMessage:
    _role: ChatRole
    _content: Sequence[ChatMessageContentT]
    _meta: Dict[str, Any] = field(default_factory=dict, hash=False)
```

The idea is that these attributes are internal.
For creation, you should use class methods `from_user`, etc.
For attribute access, several properties are available: `text`, `texts`, `tool_call`, `tool_calls`, ...

Reply:
@anakin87 I get this part, but I still hope the serialized version can drop the underscore prefixes for easier access. For example, for a deployed ChatBot pipeline, I'd prefer something like `{"chat_history": [{"role": "user", "content": [...]}, {"role": "assistant", "content": [...]}, "meta": {...}]}` in a request instead of `{"chat_history": [{"_role": "user", "_content": [...]}, {"_role": "assistant", "_content": [...]}, "_meta": {...}]}`.

Also it makes migration a bit easier.

Reply:
Also as a bonus I have something like this in my draft PR:
```python
        if isinstance(data["content"], str):
            content.append(TextContent(text=data["content"]))
        else:
            for part in data["content"]:
                ...
```
This makes the serialization form of a ChatMessage fully backwards compatible.

Reply:
@anakin87 wdyt about this? pinging in case this was lost in the threads.

Reply:
@LastRemote sorry for not responding sooner.

Since this is a substantial change, we decided to make it visible and not backward compatible.

As soon as possible, we will publish a detailed migration guide + GitHub/Discord announcements. (Tracked in #8623 and #8654).

Reply:
@anakin87 Thanks for the reply, but I'm still not fully convinced. We can still keep the internal attributes with underscore prefixes, but just for serialization/deserialization I would suggest us dropping the those for the sake of simplicity. It is still visible.

In my current use cases, I have many deployed haystack pipelines that utilize List[ChatMessage] as a part of input/output (which is very common in chatbots to track chat history and multiple reply candidates), it makes more sense to have something like
```
{"chat_history": [{"role": "user", "content": [...]}, {"role": "assistant", "content": [...]}, "meta": {...}]}
```
instead of adding an underscore prefix in all fields. Of course, this could be mitigated by some customized logic on the server end (not sure hayhook does this), but IMO that adds a layer of unneeded complexity.

We can still keep the internal attributes with underscore prefixes, but just for serialization/deserialization I would suggest us dropping the those for the sake of simplicity.